### PR TITLE
feat: evaluate typed AST

### LINKED ISSUE #208: Implement an evaluator to generate CDK code
The current implementation of deCDK is parsing and evaluating at the same time (with the exception of the cross-references between resources). Replace this strategy with the ones normally used in compilers: first parse the input and generate an abstract syntax tree and then evaluate the tree, generating the target code.

The parsing has already been implemented (actually copied from Rico's [cfnengine](https://github.com/rix0rrr/cfngine)). We now need to implement the evaluator.

### PR DESCRIPTION
**Pull request by mrgrain**
Closes #208
 
--- 
Implements an evaluator to turn the AST into a CDK app.

The aim of this PR is feature parity with the current implementation, i.e. synth tests pass without changes their snapshots.

Missing features marked as `@todo` should be added later.

--- 
For ease of review, this PR might be reviewed per commit. Especially the first (copy and paste from cfngine) and last (cleanup) make this PR noisy.

### CODE REVIEW

File: src/evaluate/evaluate.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+  private currentResource?: ResourceLike;
+
+  constructor(public readonly context: EvaluationContext) {}
+
+  public evaluateTemplate(block: <T>(fn: () => T) => T) {
+    this.context.template.resourceGraph().forEach((logicalId, resource) => {
+      block(() =>
+        this.evaluateResource(
+          logicalId,
+          resolveResourceLike(resource, logicalId, this.context.typeSystem)
```
Comment (otaviomacedo) [inline src/evaluate/evaluate.ts:33]: For everything else, type resolution is done by the parser. But resource-like is being resolved here, at the evaluation phase. Why is that?

**Review by otaviomacedo — COMMENTED** [review]
This is brilliant!

File: src/evaluate/evaluate.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+              x.transformName,
+              this.evaluateObject(x.parameters)
+            );
+          case 'and':
+            return this.fnAnd(x.operands.map(ev).map(assertBoolean));
+          case 'or':
+            return this.fnOr(x.operands.map(ev).map(assertBoolean));
+          case 'not':
+            return this.fnNot(assertBoolean(ev(x.operand)));
+          case 'equals':
```
[the remark this answers is not in the document]
Comment (mrgrain) [reply src/evaluate/evaluate.ts:125]: Yep, tracked in https://github.com/cdklabs/decdk/issues/216

**Review by otaviomacedo — APPROVED** [review]

File: src/evaluate/evaluate.ts (hunk 1 above)
Comment (mrgrain) [reply src/evaluate/evaluate.ts:33]: Good question. Short answer: No particular reason. We can move things around.

--- 

Grammar parsing, type resolution and evaluation are all distinct steps.
They all need to be initiated.

- Parsing is currently initiated here: https://github.com/cdklabs/decdk/blob/145238a53f1f9f8db5c1808500e6bac4c8a0d4a1/src/decdk.ts#L20
- Evaluation here: https://github.com/cdklabs/decdk/blob/82c309792aeb02330377c6e1382ca0eacd25eee9/src/declarative-stack.ts#L31

You are right to point out that the type resolution should be handled explicitly.

--- 

We will have to do something like this:
```
declare template: Template;
const typedTemplate: TypedTemplate = resolveTemplate(template);
new Evaluator(context).evaluateTemplate();
```

File: src/evaluate/context.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import * as cdk from 'aws-cdk-lib';
+import { IConstruct } from 'constructs';
+import * as reflect from 'jsii-reflect';
+import { Template } from '../parser/template';
+export type ContextValue = IConstruct | string | string[];
+
+export interface ContextRecord {
```
[the remark this answers is not in the document]
Comment (mrgrain) [reply src/evaluate/context.ts:7]: Fixed in #239

**Pull request merged**